lib/utils/mailup.ts

```diff
@@ -0,0 +1,287 @@
+/**
+ * Implements a Nodemailer MailUp transport.
+ *
+ * Uses the MailUp REST API to send transactional emails:
+ * see http://help.mailup.com/display/mailupapi/Transactional+Emails+using+APIs
+ *
+ */
+import * as t from "io-ts";
+import * as request from "superagent";
+
+import { Either, isLeft, isRight, left, right } from "fp-ts/lib/Either";
+import { EmailString, NonEmptyString } from "./strings";
+
+import * as nodemailer from "nodemailer";
+
+import { Address as NodemailerAddress } from "nodemailer/lib/addressparser";
+
+import * as winston from "winston";
+
+import { fromNullable, none, Option, some } from "fp-ts/lib/Option";
+import { ReadableReporter } from "./validation_reporters";
+
+// request timeout in milliseconds
+const DEFAULT_REQUEST_TIMEOUT_MS = 10000;
+
+// number of retries in case of HTTP errors
+const DEFAULT_REQUEST_RETRIES = 2;
+
+export const ENDPOINTS = {
+  sendTransactionalMail: "https://send.mailup.com/API/v2.0/messages/sendmessage"
+};
+
+type HttpMethod = "GET" | "POST";
+
+const TRANSPORT_NAME = "MailUp";
+const TRANSPORT_VERSION = "0.1";
+
+/**
+ * You need to create a SMTP+ user in MailUp administration panel
+ */
+export const SmtpAuthInfo = t.interface({
+  Secret: NonEmptyString,
+  Username: NonEmptyString
+});
+
+export type SmtpAuthInfo = t.TypeOf<typeof SmtpAuthInfo>;
+
+/**
+ * MailUp API calls common response fields
+ */
+const ApiResponse = t.interface({
+  Code: t.string,
+  Message: t.string,
+  Status: t.string
+});
+
+type ApiResponse = t.TypeOf<typeof ApiResponse>;
+
+const Address = t.interface({
+  Email: EmailString,
+  Name: t.string
+});
+
+type Address = t.TypeOf<typeof Address>;
+
+const NameValue = t.interface({
+  N: NonEmptyString,
+  V: t.string
+});
+
+type NameValue = t.TypeOf<typeof NameValue>;
+
+const Html = t.interface({
+  Body: NonEmptyString
+});
+
+type Html = t.TypeOf<typeof NameValue>;
+
+const EmailPayload = t.intersection([
+  t.interface({
+    ExtendedHeaders: t.array(NameValue),
+    From: Address,
+    Html,
+    Subject: NonEmptyString,
+    Text: NonEmptyString,
+    To: t.array(Address)
+  }),
+  t.partial({
+    Bcc: t.array(Address),
+    Cc: t.array(Address),
+    ReplyTo: t.string
+  })
+]);
+
+type EmailPayload = t.TypeOf<typeof EmailPayload>;
+
+export interface IMailUpTransportOptions {
+  readonly creds: SmtpAuthInfo;
+}
+
+interface IAddresses {
+  readonly bcc?: ReadonlyArray<NodemailerAddress>;
+  readonly cc?: ReadonlyArray<NodemailerAddress>;
+  readonly from?: ReadonlyArray<NodemailerAddress>;
+  readonly sender?: ReadonlyArray<NodemailerAddress>;
+  readonly "reply-to"?: ReadonlyArray<NodemailerAddress>;
+  readonly to?: ReadonlyArray<NodemailerAddress>;
+}
+
+function callMailUpApi(
+  method: HttpMethod,
+  url: string,
+  creds: SmtpAuthInfo,
+  payload: {}
+): Promise<Either<Error, ApiResponse>> {
+  return request(method, url)
+    .timeout(DEFAULT_REQUEST_TIMEOUT_MS)
```

Comment:
what happens to the `Promise` when the req times our or max retries are reached?

Reply:
promise is reject with error (caught in catch())

Reply:
I think that from the caller point of view, this method should be very very simple, it should either send the email successfully or fail - once we have PR #133 merged, I don't see the point for having a different retry logic hidden here

Reply:
many reasons for not having a retry logic here:

* retry logic is opaque to the caller (what errors trigger a retry? parameters can't be configured)
* there's not logging whatsoever, the call can spend up to 10s to complete and we have no clue about what's going on inside (besides we decided to rely on message rescheduling to implement the retry mechanism to avoid long running message handlers)
* [the doc](http://visionmedia.github.io/superagent/#retrying-requests) says to use the retry logic only when the request is idempotent (definitely not our case)

Reply:
Agreed and removed the retry line. Note that the retry here was harmless has it would have occurred only on timeouts and server errors, moreover most of the statements above (ie. retry opaque to caller) apply to the azure api calls as well, as they use a [retry policy](https://github.com/Azure/autorest/blob/master/docs/client/retry.md) similar to the one implemented by superagent.